### PR TITLE
Zoneout implementation for RNN units + new implementation for GRU and LSTM

### PR DESCRIPTION
**Pull request by Separius**
+Zoneout implementation for all the RNN units without breaking the API
+Another implementation for GRU and LSTM which is like implementation 0 and 2 together (it uses fused matrices for forget gate and input gates and it also computes input times recurrent kernel in a big matrix)
+using implementation=some random number is not informative so I wrote an Enum for it without breaking the old codes
+deleted some duplicate codes for generating dropout masks by defining a function

### CODE REVIEW

File: keras/layers/convolutional_recurrent.py [hunk 1]
```diff
@@ -331,6 +331,8 @@ def __init__(self, filters,
 
         self.dropout = min(1., max(0., dropout))
         self.recurrent_dropout = min(1., max(0., recurrent_dropout))
+        if 0 < (self.dropout + self.recurrent_dropout):
```
Comment (Dapid) [inline keras/layers/convolutional_recurrent.py]: I think  `if self.dropout > 0 or self.recurrent_dropout > 0` is more clear.

File: keras/layers/convolutional_recurrent.py [hunk 2]
```diff
@@ -494,7 +496,7 @@ def input_conv(self, x, w, b=None, padding='valid'):
                                   data_format=self.data_format)
         return conv_out
 
-    def reccurent_conv(self, x, w):
+    def recurrent_conv(self, x, w):
```
Comment (Dapid) [inline keras/layers/convolutional_recurrent.py]: `reccurent_conv` was a typo, but it is now part of the API. Keep it as an alias for `recurrent_conv` (with deprecation warning?)

File: keras/layers/recurrent.py [hunk 3]
```diff
@@ -123,6 +161,11 @@ class Recurrent(Layer):
             enabling more time-efficient parallelization on the GPU.
             Note: RNN dropout must be shared for all gates,
             resulting in a slightly reduced regularization.
+            If set to 3 (LSTM/GRU only),
+            not only the RNN combines the gates into a single matrix,
+            but it also uses large matrices for input
+            Note: RNN dropout must be shared for all gates,
+            resulting in a slightly reduced regularization.
```
Comment (Dapid) [inline keras/layers/recurrent.py]: Can you add a reference paper to this?

File: keras/layers/recurrent.py [hunk 4]
```diff
@@ -59,6 +67,36 @@ def _time_distributed_dense(x, w, b=None, dropout=None,
...
+    if 0 < zoneout_probability < 1:
+        return new_state * mask + old_state * (1. - mask)
+    else:
+        return new_state
+
+
+def _gen_dropout_mask(probability, inputs, units, training, repeats=1, is_zoneout=False):
+    if 0 < probability < 1:
+        ones = K.ones_like(K.reshape(inputs[:, 0, 0], (-1, 1)))
+        ones = K.tile(ones, (1, units))
```
Comment (Dapid) [inline keras/layers/recurrent.py]: Can you make it in one step? Like (untested):

    ones = K.ones((inputs._keras_shape[0], units))

File: keras/layers/convolutional_recurrent.py (hunk 1 above)
Comment (Separius) [inline keras/layers/convolutional_recurrent.py]: I agree, but I followed the old code, I mean the code from Recurrent class (in the call function if I remember correctly)
but if you say so, I'll change it

File: keras/layers/convolutional_recurrent.py (hunk 2 above)
Comment (Separius) [inline keras/layers/convolutional_recurrent.py]: yeah, sorry for breaking the API, I'll make an alias for it

File: keras/layers/recurrent.py (hunk 4 above)
Comment (Separius) [inline keras/layers/recurrent.py]: if I wanna be honest, I don't know how to do it (I am terrible at shapes), but I'll test your solution

File: keras/layers/recurrent.py (hunk 3 above)
Comment (Separius) [inline keras/layers/recurrent.py]: no, I think there is no reference to this and it's not a big deal actually
it "cloud be" beneficial on some hardwares, which are good at multiplying big and medium matrices

**Comment by Separius** [comment]
Here's the reference to the paper: https://arxiv.org/abs/1606.01305v3
for testing, I used this for my own problem, which is emoji prediction and it is much better than the recurrent_dropout (I'll open source it as soon as possible)
I also changed the rnn example on imdb, and the zoneout improved the vanilla RNN and the GRU, but actually decreased the accuracy with LSTM and I think it's duo to the fact that there are two hyper parameters there and not just one
ok, I will rebase it

**Comment by Separius** [comment]
@Dapid did you read the reference?

**Comment by Separius** [comment]
I was going to write batch norm for RNN and read an old PR for it, so I realized that my zoneout implementation is bloating the code, and this way of implementing it is wrong.
There might be a much cleaner way of doing this with layer wrappers and I think that's the way to go, so I close my current PR and if I wrote the wrapper version, I'll make another PR (or reopen this one)
